Add in call backs for drag/resize calls

browniefed opened the pull request:
Thoughts on adding callbacks for this stuff?
I just basically passed through most information since I wasn't sure exactly what people were going to want to utilize.

STRML commented on build/GridItem.js:
```diff
@@ -70,7 +70,7 @@ var GridItem = React.createClass({
   },
 
-  getDefaultProps: function () {
+  getDefaultProps: function getDefaultProps() {
```
Which version of 6to5 are you using? I wonder why the build came out this way for you; this is definitely better for stack traces.

browniefed replied:
I just did an npm install so whatever the package.json says

STRML replied:
Gotcha. I had an old version in npm cache. Could you rebase to master so we can squash the diffs?

browniefed replied:
Will do, I'm at work now so I'll handle all these sometime later tonight.

STRML replied:
Cool. Thanks a lot for the contribution, I appreciate it.
On 1/8/15 6:50 PM, Jason Brown wrote:
>
> In build/GridItem.js
> <https://github.com/STRML/react-grid-layout/pull/12#discussion-diff-22667656>:
>
> > @@ -70,7 +70,7 @@ var GridItem = React.createClass({
> >      handle: React.PropTypes.string
> >    },
> >  
> > -  getDefaultProps: function () {
> > +  getDefaultProps: function getDefaultProps() {
>
> Will do, I'm at work now so I'll handle all these sometime later tonight.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/STRML/react-grid-layout/pull/12/files#r22667656>.
>

STRML commented on lib/ReactGridLayout.jsx:
```diff
@@ -57,7 +57,16 @@ var ReactGridLayout = React.createClass({
+    // Cals when drag is complete
+    onDragStop: React.PropTypes.func,
+    // Calls when resize starts, allows to cancel/restrict resize
+    onResize: React.PropTypes.func,
```
I suppose we should add an onResizeStart to be consistent?

browniefed replied:
I'll add it

STRML commented on lib/ReactGridLayout.jsx:
```diff
@@ -152,11 +166,33 @@ var ReactGridLayout = React.createClass({
+  onDragStart(i, x, y, {e, element, position}) {
+    var layout = this.state.layout;
+    var l = utils.getLayoutItem(layout, i);
+    if (this.props.onDragStart(layout, this.state.layout, l, null, {e, element, position}) === false) {
```
I like making it cancelable via the callback; any ideas how this is done in other libraries? My instinct is to check if the event has been canceled (e.g. e.preventDefault()) but that doesn't quite make as much sense in React.

STRML replied:
Of course if we did that we'd have to be consistent among all the other callbacks as well. With this code as-is the placeholder would quickly be replaced by the drag handler below.

browniefed replied:
Yeah, I think e.preventDefault would be the way to go since I'm not necessarily a fan of returning false, and as of React .12 you must call preventDefault to stop.
It looks like we could do it with mouseDown event since that is what react-draggable is using, however we may need to mod the handleStartDrag to actually do the preventDefault check.

STRML replied:
That's fine by me, I'll add that to my fork of draggable.

browniefed commented:
Rebased, added onResizeStart, add README documentation.
Only thing this will depend on is react-draggable and react-resizable stopping drag/resize when `preventDefault` gets called.

browniefed commented:
Hey just curious if you had an idea of when these could get merged?

STRML commented:
Hey Jason - sorry, chasing down a lot of things on a lot of projects
right now. Should be merged soon, in the meantime I would recommend
keeping this rebased on master and using a github url in your project
until I can finish reviewing.
On 1/14/15 5:36 PM, Jason Brown wrote:
> Hey just curious if you had an idea of when these could get merged?
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/STRML/react-grid-layout/pull/12#issuecomment-69945453

browniefed commented:
@STRML no problem, thanks

STRML commented on lib/ReactGridLayout.jsx:
```diff
@@ -201,13 +258,20 @@ var ReactGridLayout = React.createClass({
       w: w, h: h, x: l.x, y: l.y, placeholder: true, i: i
     };
+
+    activeDrag = this.props.onResize(layout, layout, l, activeDrag, {e, element, size}) || activeDrag;
```
In general I don't really like the assignment here; I don't see any demand for it and it is confusing - on dragStart and resizeStart, returning a value does nothing, on drag, you return the layout, on resize, you return the resizing item, and on dragStop and resizeStop you return the layout again.

Do you have a particular need for that functionality? Otherwise let's just remove the assignment altogether.

browniefed replied:
Well onResizeStart/onDragStart you ideally would be canceling with preventDefault so no return necessary.
Ideally if there were a way to set both full layout/active drag layout items that would be ideal.
I'm attempting to implement a lot of restrictions around dragging/resizing  so controlling the full layout/actively dragging/resizing item is what I need.

I realize I can modify the layout and activeDrag by just simple object reference but that's just bad. Besides calling a setState of ReactGridLayout and passing down new props layout during the drag (which would probably be very inefficient and cause some weird UI states temporarily) I'm not sure what else is the solution here.

Basically how would you allow the user to control the positioning of the entire layout/actively dragging item while it's happening?

STRML commented on README.md:
```diff
@@ -181,6 +181,18 @@ listenToWindowResize: React.PropTypes.bool,
+onDrag: React.PropTypes.func,
+// Calls when drag is complete
+onDragStop: React.PropTypes.func,
+//Calls when resize starts, allows for canceling
```
Could you please fix the spacing here and note the callback signature on each one both here and in the source?

For example: 

"Calls when resize starts. Callback is of the signature `(newLayout, oldLayout, layoutItem, {e, element, size}`."

Modify the callback signature to `{e, element, position}` for drags.

browniefed commented:
@STRML hey I just decided to remove the assignment stuff for now, so I've rebased and squashed, fixed the documentation text. I think it should be good to go.

STRML commented:
Perfect. This is simple, we can worry about the assignment stuff later on.

STRML commented:
`0.7.1`
